text/0000-canonical-metadata.md

```diff
@@ -0,0 +1,1198 @@
+interface ScriptFile {
+    fileType: 'js';
+    fileName: string;
+
+    // A list of ModuleReference objects referenced by this script file
+    moduleReferences: ModuleReference[];
+
+    // List of class metadata, covers only component class and any exported class.
+    // class has to be created using the ES6 'class' syntax
+    classes?: Class[];
+
+    // Static import statements in a javascript module.
+    imports?: Import[];
+
+    // Dynamic import statements in a javascript module, additionally information about hints if provided.
+    dynamicImports?: DynamicImport[];
+
+    // Exports of local bindings
+    exports: Export[];
+    // Reexporting bindings from another module.
+    // e.g. export * from …;
+    reExports: ReExport[];
+
+    // DOM Event objects instantiated in a module. Includes Events and CustomEvents.
+    domEvents?: DOMEvent[];
+
+    // Programmatically added event listeners in a module
+    eventListeners?: ProgrammaticEventListener[];
+
+    // EventTarget.dispatchEvent invocations
+    eventsDispatched?: EventDispatch[];
+
+    /**
+     * Static resources referenced in a module.
+     * Includes any fully qualified urls referenced in javascript.
+     * Does not include resources loaded via lightning/platformResourceLoader or a url determined at runtime.
+     */
+    staticResources?: StaticResourceReference[];
+}
+
+// Information about a class
+interface Class {
+    // Is optional because class can be anonymous
+    name?: string;
+
+    // A class is considered a component class if it extends
+    // 'LightningElement' class available in the standard 'lwc' module.
+    isComponentClass?: boolean;
+
+    // A parent can be another class declared locally, imported identifier from an external module.
+    // When not statically analyzable, will be 'unresolved'(e.g. Mixin)
+    extends: ParentClass;
+
+    properties?: ClassProperty[];
+
+    methods?: ClassMethod[];
+
+    // Location starts with the "class" key word and ends at the closing curly braces of the class body
+    location: SourceLocation;
+
+    // A string literal representing the leading comments of the class declaration
+    doc?: string;
+    id: ID;
+}
+
+interface ClassMember {
+    // Note: getter/setter types are considered a property
+    type: 'property' | 'method';
+    // According to javascript specs
+    propertyFieldType: 'public' | 'private' | 'static';
+    name: string;
+    // Location starts with the first character of the member's identifier and ends at the semi-colon
+    // in case of a property and closing curly in case of a method
+    // For a accessor property type, the getter and setter property has the specific location
+    location?: SourceLocation;
+    // A string literal representing the leading comments of the class member
+    doc?: string;
+}
+
+// Information about a class property of a component class.
+interface ClassProperty extends ClassMember {
```

Comment:
Is there a proposed way of indicating `attribute` vs `property` when exposed via `buildCustomElementConstrutor`?